File: src/ex_calculator_py/ex_calculator_py/arithmetic/argument.py
import random

from ros_study_msgs.msg import ArithmeticArgument
from rcl_interfaces.msg import SetParametersResult
import rclpy
from rclpy.node import Node
from rclpy.parameter import Parameter
from rclpy.qos import QoSDurabilityPolicy
from rclpy.qos import QoSHistoryPolicy
from rclpy.qos import QoSProfile
from rclpy.qos import QoSReliabilityPolicy
from rclpy.logging import LoggingSeverity #로그 레벌 설정해주는 함수가 들어있는 라이브러리

class Argument(Node):

    def __init__(self):
        super().__init__('argument')
        self.get_logger().set_level(LoggingSeverity.INFO) #로그 레벌 설정 함수
        self.get_logger().info('Argument init set start')

        self.declare_parameter('qos_depth', 10)
        qos_depth = self.get_parameter('qos_depth').value #10이 들어간다
        self.get_logger().info(f'qos_depth set {qos_depth}') #qos_depth에 뭐가 들어갔는지 로그로 확인가능

        self.declare_parameter('min_random_num', 0)
        self.min_random_num = self.get_parameter('min_random_num').value
        self.get_logger().info(f'min_random_num set {self.min_random_num}')

        self.declare_parameter('max_random_num', 9)
        self.max_random_num = self.get_parameter('max_random_num').value
        self.get_logger().info(f'max_random_num set {self.max_random_num}')

        self.add_on_set_parameters_callback(self.update_parameter)
        self.get_logger().info(f'add_on_set_parameters_callback on')

        QOS_RKL10V = QoSProfile(
            reliability=QoSReliabilityPolicy.RELIABLE,
            history=QoSHistoryPolicy.KEEP_LAST,
            depth=qos_depth,
            durability=QoSDurabilityPolicy.VOLATILE)
        self.get_logger().info('QOS_RKL10V get QoSProfile')

        self.arithmetic_argument_publisher = self.create_publisher(
            ArithmeticArgument,
            'arithmetic_argument',
            QOS_RKL10V)
        self.timer = self.create_timer(1.0, self.publish_random_arithmetic_arguments)
        self.get_logger().info('publish topic : arithmetic_argument_publisher')
        self.get_logger().info('end init')

    def publish_random_arithmetic_arguments(self):
        self.get_logger().info('start publish_random_arithmetic_arguments')
        msg = ArithmeticArgument()
        msg.stamp = self.get_clock().now().to_msg()
        msg.argument_a = float(random.randint(self.min_random_num, self.max_random_num))
        msg.argument_b = float(random.randint(self.min_random_num, self.max_random_num))
        self.arithmetic_argument_publisher.publish(msg)
        self.get_logger().info(f'Published argument a: {msg.argument_a}')
        self.get_logger().info(f'Published argument b: {msg.argument_b}')

    def update_parameter(self, params):
        self.get_logger().info('start update_parameter')
        for param in params:
            if param.name == 'min_random_num' and param.type_ == Parameter.Type.INTEGER:
                self.min_random_num = param.value
                self.get_logger().info(f'self.min_random_num = {self.min_random_num}')
            elif param.name == 'max_random_num' and param.type_ == Parameter.Type.INTEGER:
                self.max_random_num = param.value
                self.get_logger().info(f'self.max_random_num = {self.max_random_num}')
        return SetParametersResult(successful=True)
        #여기에 로그 남기면 안됨 return에서 함수 끝

def main(args=None):
    rclpy.init(args=args)
    try:
        argument = Argument()
        try:
            rclpy.spin(argument)
        except KeyboardInterrupt:
            argument.get_logger().info('Keyboard Interrupt (SIGINT)')
        finally:
            argument.destroy_node()
    finally:
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/cpp_excalculator/src/arithmetic/argument.cpp
#include "rclcpp/rclcpp.hpp"
#include "rcl_interfaces/msg/set_parameters_result.hpp"
#include "ros_study_msgs/msg/arithmetic_argument.hpp"

#include <random>

class Argument : public rclcpp::Node
{
public:
    Argument() : Node("argument")
    {
        RCLCPP_INFO(this->get_logger(), "Argument init set start");

        // Declare parameters
        //this는 Argument의 포인터
        this->declare_parameter("qos_depth", 10); 
        this->declare_parameter("min_random_num", 0);
        this->declare_parameter("max_random_num", 9);

        qos_depth_ = this->get_parameter("qos_depth").as_int();
        min_random_num_ = this->get_parameter("min_random_num").as_int();
        max_random_num_ = this->get_parameter("max_random_num").as_int();

        RCLCPP_INFO(this->get_logger(), "qos_depth set %d", qos_depth_);
        RCLCPP_INFO(this->get_logger(), "min_random_num set %d", min_random_num_);
        RCLCPP_INFO(this->get_logger(), "max_random_num set %d", max_random_num_);

        // Parameter callback
        parameter_callback_handle_ = this->add_on_set_parameters_callback(
            std::bind(&Argument::on_parameter_changed, this, std::placeholders::_1));

        RCLCPP_INFO(this->get_logger(), "add_on_set_parameters_callback on");

        // QoS 설정
        rclcpp::QoS qos(rclcpp::QoSInitialization::from_rmw(rmw_qos_profile_default));
        qos.keep_last(qos_depth_);
        qos.reliability(RMW_QOS_POLICY_RELIABILITY_RELIABLE);
        qos.durability(RMW_QOS_POLICY_DURABILITY_VOLATILE);

        // Publisher 생성
        publisher_ = this->create_publisher<ros_study_msgs::msg::ArithmeticArgument>(
            "arithmetic_argument", qos);

        timer_ = this->create_wall_timer(
            std::chrono::seconds(1),
            std::bind(&Argument::publish_random_message, this));

        RCLCPP_INFO(this->get_logger(), "publish topic : arithmetic_argument_publisher");
        RCLCPP_INFO(this->get_logger(), "end init");
    }

private:

rclcpp::Publisher<ros_study_msgs::msg::ArithmeticArgument>::SharedPtr publisher_;
rclcpp::TimerBase::SharedPtr timer_;
OnSetParametersCallbackHandle::SharedPtr parameter_callback_handle_;

int qos_depth_;
int min_random_num_;
int max_random_num_;

    void publish_random_message()
    {
        ros_study_msgs::msg::ArithmeticArgument msg;
        msg.stamp = this->now();

        msg.argument_a = static_cast<float>(get_random(min_random_num_, max_random_num_));
        msg.argument_b = static_cast<float>(get_random(min_random_num_, max_random_num_));

        publisher_->publish(msg);

        RCLCPP_INFO(this->get_logger(), "Published argument a: %.1f", msg.argument_a);
        RCLCPP_INFO(this->get_logger(), "Published argument b: %.1f", msg.argument_b);
    }

    rcl_interfaces::msg::SetParametersResult on_parameter_changed(
        const std::vector<rclcpp::Parameter> &params)
    {
        RCLCPP_INFO(this->get_logger(), "start update_parameter");

        for (const auto &param : params)
        {
            if (param.get_name() == "min_random_num" && param.get_type() == rclcpp::ParameterType::PARAMETER_INTEGER)
            {
                min_random_num_ = param.as_int();
                RCLCPP_INFO(this->get_logger(), "self.min_random_num = %d", min_random_num_);
            }
            else if (param.get_name() == "max_random_num" && param.get_type() == rclcpp::ParameterType::PARAMETER_INTEGER)
            {
                max_random_num_ = param.as_int();
                RCLCPP_INFO(this->get_logger(), "self.max_random_num = %d", max_random_num_);
            }
        }

        rcl_interfaces::msg::SetParametersResult result;
        result.successful = true;
        return result;
    }

    int get_random(int min, int max)
    {
        static std::random_device rd;
        static std::mt19937 gen(rd());
        std::uniform_int_distribution<> dis(min, max);
        return dis(gen);
    }

};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<Argument>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
